closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -469,39 +475,50 @@ int decryptFinal(byte[] in, int inOfs, int len,
                      byte[] out, int outOfs)
         throws IllegalBlockSizeException, AEADBadTagException,
         ShortBufferException {
-        if (len < 0) {
-            throw new ProviderException("Input length is negative");
-        }
-        if (len < tagLenBytes - ibuffer.size()) {
-            throw new AEADBadTagException("Input too short - need tag");
-        }
-        if (len > MAX_BUF_SIZE - ibuffer.size()) {
-            throw new ProviderException("SunJCE provider only supports "
-                + "a positive input size up to " + MAX_BUF_SIZE + " bytes");
-        }
 
-        if (out.length - outOfs < len  + ibuffer.size() - tagLenBytes) {
-            throw new ShortBufferException("Output buffer too small");
-        }
+        byte[] aad;
+        // avoid any race conditions in OpenSSL when the key/iv are used concurrently
+        byte[] keyCopy = key.clone();
+        byte[] ivCopy = iv.clone();
 
-        byte[] aad = ((aadBuffer == null || aadBuffer.size() == 0) ?
-                       emptyAAD : aadBuffer.toByteArray());
 
-        aadBuffer = null;
+        // need to synchronize this block as if there are concurrent calls to decryptFinal
+        // the shared variables aadBuffer and ibuffer may corrupt input and cause a segfault
+        // in openssl code
+        synchronized (this) {
+            if (len < 0) {
+                throw new ProviderException("Input length is negative");
+            }
 
-        if (len > 0) {
-            ibuffer.write(in, inOfs, len);
-        }
+            if (len < (tagLenBytes - ibuffer.size())) {
+                throw new AEADBadTagException("Input too short - need tag");
+            }
+
+            if (len > (MAX_BUF_SIZE - ibuffer.size())) {
+                throw new ProviderException("SunJCE provider only supports "
+                    + "a positive input size up to " + MAX_BUF_SIZE + " bytes");
+            }
 
-        // refresh 'in' to all buffered-up bytes
-        in = ibuffer.toByteArray();
-        inOfs = 0;
-        len = in.length;
-        ibuffer.reset();
+            if ((out.length - outOfs) < (len  + ibuffer.size() - tagLenBytes)) {
+                throw new ShortBufferException("Output buffer too small");
+            }
+
+            aad = ((aadBuffer == null) || (aadBuffer.size() == 0)) ? emptyAAD : aadBuffer.toByteArray();
```

Comment:
Thanks the clarification.

Got a thought here, the `ByteArrayOutputStream` methods used are all `synchronized`.
However `aadBuffer` is used by both `encryptFinal` and `decryptFinal`. Can each method have its own dedicated `aadBuffer` and hopefully avoid synchronization?
Just out of curiosity, is there any perf benchmark for native crypto?

Reply:
aadBuffer can't be split into one buffer for encrypt and one for decrypt as the AAD bytes are provided via updateAAD() call which doesn't tell you if the bytes are destined for encrypt or decrypt, so you need to keep it in a shared buffer until you're at the encrypt/decrypt call.

you can find perf. benchmarks for crypto (native or not) at https://github.com/openjdk/jmh-jdk-microbenchmarks

Reply:
Okay, in this case, is the `aadBuffer` or `ibuffer_enc` related method required to be synchronized? Do we happen to know the failing scenario with multiple threads doing `encrypt`, `decrypt` or combination of both?

Reply:
the failing scenario was doing concurrent decrypts. there may be a way to fix the crash without a synchronize, i'll work on it.